metricbeat/module/kibana/cluster_actions/cluster_actions.go

```diff
@@ -0,0 +1,111 @@
+// Licensed to Elasticsearch B.V. under one or more contributor
+// license agreements. See the NOTICE file distributed with
+// this work for additional information regarding copyright
+// ownership. Elasticsearch B.V. licenses this file to you under
+// the Apache License, Version 2.0 (the "License"); you may
+// not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing,
+// software distributed under the License is distributed on an
+// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+
+package cluster_actions
+
+import (
+	"fmt"
+
+	"github.com/pkg/errors"
+
+	"github.com/elastic/beats/v7/metricbeat/helper"
+	"github.com/elastic/beats/v7/metricbeat/mb"
+	"github.com/elastic/beats/v7/metricbeat/mb/parse"
+	"github.com/elastic/beats/v7/metricbeat/module/kibana"
+)
+
+// init registers the MetricSet with the central registry.
+// The New method will be called after the setup of the module and before starting to fetch data
+func init() {
+	mb.Registry.MustAddMetricSet(kibana.ModuleName, "cluster_actions", New,
+		mb.WithHostParser(hostParser),
+	)
+}
+
+var (
+	hostParser = parse.URLHostParserBuilder{
+		DefaultScheme: "http",
+		DefaultPath:   kibana.ClusterActionsPath,
+	}.Build()
+)
+
+// MetricSet type defines all fields of the MetricSet
+type MetricSet struct {
+	*kibana.MetricSet
+	actionsHTTP *helper.HTTP
+}
+
+// New create a new instance of the MetricSet
+func New(base mb.BaseMetricSet) (mb.MetricSet, error) {
+	ms, err := kibana.NewMetricSet(base)
+	if err != nil {
+		return nil, err
+	}
+
+	return &MetricSet{
+		MetricSet: ms,
+	}, nil
+}
+
+// Fetch methods implements the data gathering and data conversion to the right format
+// It returns the event which is then forward to the output. In case of an error, a
+// descriptive error must be returned.
+func (m *MetricSet) Fetch(r mb.ReporterV2) (err error) {
+	if err = m.init(); err != nil {
+		return err
+	}
+
+	if err = m.fetchMetrics(r); err != nil {
+		return errors.Wrap(err, "error trying to get cluster action data from Kibana")
+	}
+
+	return
+}
+
+func (m *MetricSet) init() error {
```

Comment:
So, I'm not too familiar with the kibana APIs, but I'm wondering how much of this should be done from `Fetch` and how much can be done from `New()`. Does the Version call and HTTP init really need to happen every time we update the metrics?

Reply:
Definitely open to optimizing this - I don't really know the lifecycle and I mainly copy/pasted how the other modules work. What would you recommend?

Reply:
Something should only go in `Fetch()` if we actually need to call it every time the metrics are getting updated. Connection handlers, auth tokens, etc, should be set up in `New()` and kept around where possible. I'm guessing that most of the setup done in `init()` isn't expected to change over the course of a metricbeat reporting interval, and should probably be moved into `New()`.

Reply:
Okay, makes sense, thanks!